src/transform/src/predicate_pushdown.rs

```diff
@@ -607,6 +619,69 @@ impl PredicatePushdown {
         }
     }
 
+    /// If `s` is of the form
+    /// `(isnull(expr1) && isnull(expr2)) || (expr1 = expr2)`,
+    /// extract `expr1` and `expr2`.
+    fn extract_equal_or_both_null(s: &mut MirScalarExpr) -> Option<(MirScalarExpr, MirScalarExpr)> {
```

Comment:
In this branch I see a Filter being applied twice, which is not the case with the commit just prior:

```
DROP SCHEMA public CASCADE;
CREATE SCHEMA public;

CREATE TABLE t1 (f1 INTEGER, f2 INTEGER NOT NULL);
CREATE TABLE t2 (f1 INTEGER, f2 INTEGER NOT NULL);

EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON ( TRUE ) WHERE t1.f2 < t1.f1 AND 7 IS NOT NULL OR t2.f1 = 9;
```
Produces this plan:

```
materialize=> EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON ( TRUE ) WHERE t1.f2 < t1.f1 AND 7 IS NOT NULL OR t2.f1 = 9;
                  Optimized Plan                  
--------------------------------------------------
 %0 =                                            +
 | Get materialize.public.t1 (u45)               +
 | ArrangeBy ()                                  +
                                                 +
 %1 =                                            +
 | Get materialize.public.t2 (u47)               +
                                                 +
 %2 = Let l0 =                                   +
 | Join %0 %1                                    +
 | | implementation = Differential %1 %0.()      +
 | | demand = (#0..#3)                           +
 | Filter ((#2 = 9) || (#1 < #0))                + <- filter applied onde
                                                 +
 %3 =                                            +
 | Get %2 (l0)                                   +
 | Filter ((#1 < #0) || (#2 = 9))                + <- Filter applied twice
                                                 +
```

Reply:
This is actually a wrong result situation for this query:

```
SELECT  5  FROM t2  AS a1  RIGHT  JOIN t1  AS a2  ON a1 .f1  IS  NULL  WHERE a1 .f1  > a2 .f1  OR a2 .f1  >  3  AND a1 .f2  IS  NULL   ;
```
The extra filter prevents the query from returning rows that it should. The schema and data is available in a comment further down in this review.